File: packages/sdk-csharp/Melaya/RunnerApi.cs
namespace Melaya;

/// <summary>
/// Runner API — mint, list, and revoke runner tokens.
/// Maps to <c>/api/v1/private/runner/tokens</c>.
/// Runner tokens (<c>mel_run_</c> prefix) authenticate the runner CLI process.
/// </summary>
/// <example>
/// <code>
/// var result = await m.Runner.CreateTokenAsync(new RunnerTokenCreateRequest { Label = "prod-server-1" });
/// // Store result.Token securely — it is shown only once.
///
/// var tokens = await m.Runner.ListTokensAsync();
/// await m.Runner.RevokeTokenAsync(tokens[0].Id!);
/// </code>
/// </example>
public sealed class RunnerApi
{
    private readonly MelayaHttpClient _http;

    internal RunnerApi(MelayaHttpClient http) => _http = http;

    /// <summary>
    /// Mint a new runner token.
    /// The plaintext token is returned only in this response — store it securely.
    /// </summary>
    public async Task<RunnerTokenCreateResult> CreateTokenAsync(RunnerTokenCreateRequest? body = null, CancellationToken ct = default)
    {
        return await _http.PostAsync<RunnerTokenCreateResult>("/api/v1/private/runner/tokens", body, ct).ConfigureAwait(false);
    }

    /// <summary>List all runner tokens for the caller (masked, with last_seen).</summary>
    public async Task<List<RunnerToken>> ListTokensAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<RunnerToken>>("/api/v1/private/runner/tokens", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Revoke a runner token by ID.</summary>
    public async Task<BoolResult> RevokeTokenAsync(string tokenId, CancellationToken ct = default)
    {
        return await _http.DeleteAsync<BoolResult>(
            $"/api/v1/private/runner/tokens/{Uri.EscapeDataString(tokenId)}", ct: ct).ConfigureAwait(false);
    }
}

File: packages/sdk-kotlin/src/main/kotlin/org/melaya/RunnerAPI.kt
package org.melaya

import org.json.JSONObject

/**
 * Runner API — mint, list, and revoke runner tokens (`mel_run_` prefix).
 *
 * Runner tokens authenticate the Melaya runner CLI process that executes
 * agent pipelines on your infrastructure.
 *
 * Paths:
 *   - `POST   /api/v1/private/runner/tokens`           — mint a new token
 *   - `GET    /api/v1/private/runner/tokens`           — list all tokens (masked)
 *   - `DELETE /api/v1/private/runner/tokens/:tokenId`  — revoke a token
 *
 * @example
 * ```kotlin
 * val result = melaya.runner.createToken(label = "prod-server-1")
 * // result.getString("token") — store securely, shown only once
 * val tokens = melaya.runner.listTokens()
 * melaya.runner.revokeToken(tokens[0].getString("id"))
 * ```
 */
class RunnerAPI internal constructor(private val http: HttpClient) {

    /**
     * Mint a new runner token.
     * The plaintext token is returned only in this response — store it securely.
     */
    fun createToken(label: String? = null): JSONObject {
        val body = buildMap<String, Any?> {
            if (label != null) put("label", label)
        }
        return http.post("/api/v1/private/runner/tokens", body).asObject()
    }

    /** List all runner tokens for the caller (masked, with last_seen timestamp). */
    fun listTokens(): List<JSONObject> {
        val r = http.get("/api/v1/private/runner/tokens")
        return when (r) {
            is org.json.JSONArray -> r.toJsonObjects()
            is JSONObject -> r.optJSONArray("tokens")?.toJsonObjects() ?: emptyList()
            else -> emptyList()
        }
    }

    /** Revoke a runner token by its [tokenId]. */
    fun revokeToken(tokenId: String): JSONObject {
        return http.delete("/api/v1/private/runner/tokens/${enc(tokenId)}").asObject()
    }

    private fun enc(s: String) = java.net.URLEncoder.encode(s, "UTF-8")
}
